confluent_kafka/producer.py

```diff
@@ -0,0 +1,139 @@
+        Create a new Kafka Producer instance with or without serializer support.
+
+        To avoid spontaneous calls from non-Python threads all callbacks will only be served upon
+            calling ```client.poll()``` or ```client.flush()```.
+
+        :param dict conf: Configuration properties. At a minimum ``bootstrap.servers`` **should** be set.
+            See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md for more information.
+        :param func key_serializer(topic, key): Converts key to bytes.
+            **note** serializers are responsible for handling NULL keys
+        :param func value_serializer(topic, value): Converts value to bytes.
+            **note** serializers are responsible for handling NULL values
+        :param func error_cb(kafka.KafkaError): Callback for generic/global error events.
+        :param func stats_cb(json_str): Callback for statistics emitted every ``statistics.interval.ms``.
+            See https://github.com/edenhill/librdkafka/wiki/Statistics” for more information.
+        :param func throttle_cb(confluent_kafka.ThrottleEvent): Callback for throttled request reporting.
+        :param logging.handlers logger: Forwards logs from the Kafka client to the provided handler instance.
+            Log messages will only be forwarded when ``client.poll()`` or ``producer.flush()`` are called.
+        :raises TypeError: If conf is not a dict object.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if 'key_serializer' in kwargs or 'value_serializer' in kwargs:
+            return super(Producer, cls).__new__(SerializingProducer, *args, **kwargs)
+        return super(Producer, cls).__new__(cls, *args, **kwargs)
+
+
+class SerializingProducer(Producer):
+    """
+        SerializingProducer extends Producer with configurable key and value serializer.
+
+        To avoid spontaneous calls from non-Python threads all callbacks will only be served upon
+            calling ```client.poll()``` or ```client.flush()```.
+
+        Instances of SerializingProducer cannot be created directly.
+        To obtain an instance of this class instantiate a Consumer with a key and/or value deserializer.
+
+        Duplicate params have been omitted for brevity. See Consumer for class documentation.
+
+        :raises TypeError: If conf is not a dict.
+        :raises TypeError: If instantiated directly.
+    """
+
+    __slots__ = ["_key_serializer", "_value_serializer"]
+
+    def __new__(cls, *args, **kwargs):
+        raise TypeError("SerializingProducer is a non user-instantiable class")
+
+    @staticmethod
+    def byteSerializer(topic, data):
+        """ Pass-through serializer """
+        return data
+
+    # conf must remain optional as long as kwargs are supported
+    def __init__(self, conf={}, key_serializer=None, value_serializer=None,
+                 error_cb=None, stats_cb=None, throttle_cb=None, logger=None, **kwargs):
+
+        if not isinstance(conf, dict):
```

Comment:
same comments here on type checking arguments and prefer not to have the fancy `__new__` (depending on perf benefit).

Reply:
ack, I'm partial to checking the type over try except personally but I don't have terribly strong feelings either way.